lib/galaxy/tools/parameters/__init__.py

```diff
@@ -312,10 +312,90 @@ def populate_state(request_context, inputs, incoming, state, errors={}, prefix='
     3
     >>> print(state['b'][0]['d'][0]['f']['h'])
     4
+    >>> # now test with input_format='21.01'
+    >>> flat_new = {'a': 1, 'b': [{'c': 2, 'd': [{'e': 3, 'f': {'h': 4, 'g': True}}]}]}
+    >>> state_new = OrderedDict()
+    >>> populate_state(trans, inputs, flat_new, state_new, check=False, input_format='21.01')
+    >>> print(state_new['a'])
+    1
+    >>> print(state_new['b'][0]['c'])
+    2
+    >>> print(state_new['b'][0]['d'][0]['e'])
+    3
+    >>> print(state_new['b'][0]['d'][0]['f']['h'])
+    4
+
     """
+    if input_format == 'legacy':
+        prefix = ''
+        _populate_state_legacy(request_context, inputs, incoming, state, errors, prefix, context, check, simple_errors)
+        return
+    elif input_format == '21.01':
+        context = ExpressionContext(state, context)
+        for input in inputs.values():
+            state[input.name] = input.get_initial_value(request_context, context)
+            group_state = state[input.name]
+            if input.type == 'repeat':
+                if len(incoming[input.name]) > input.max or len(incoming[input.name]) < input.min:
+                    errors[input.name] = 'The number of repeat elements is outside the range specified by the tool.'
+                else:
+                    del group_state[:]
+                    for rep_index, rep in enumerate(incoming[input.name]):
+                        new_state = {'__index__' : rep_index}
```

Comment:
We should be able to drop `__index__`. This was for backwards compatibility (8 years ago!), according to https://github.com/mvdbeek/galaxy/blob/cdad7c79a9826bbb9273d5776cdcb264da8e0e1e/lib/galaxy/tools/parameters/grouping.py#L111